File: app/src/main/java/hu/kts/cmetronome/ui/main/IndicatorAnimation.java
package hu.kts.cmetronome.ui.main;

import android.animation.Animator;
import android.animation.ObjectAnimator;
import android.app.Activity;
import android.content.res.Resources;
import android.view.View;

import butterknife.BindView;
import butterknife.ButterKnife;
import hu.kts.cmetronome.R;
import hu.kts.cmetronome.functional.Consumer;

/**
 * Created by andrasnemeth on 12/01/16.
 *
 * AnimatorSet is avoided because it behaved strange on cancel.
 */
public class IndicatorAnimation {

    @BindView(R.id.metronome)
    View metronomeView;
    @BindView(R.id.metronome_indicator)
    View indicatorView;

    private Consumer<Event> callback;
    private boolean animationRunning;
    private ObjectAnimator down;
    private ObjectAnimator right;
    private ObjectAnimator up;
    private ObjectAnimator left;
    private ObjectAnimator currentAnimation;

    private Animator.AnimatorListener animatorListener = new Animator.AnimatorListener() {
        @Override
        public void onAnimationStart(Animator animation) {
            invokeStartCallback(animation);
        }

        @Override
        public void onAnimationEnd(Animator animation) {
            if (animationRunning) {
                currentAnimation = getNextAnimation(animation);
                if (isLastPartOfTheCycle(animation)) {
                    callback.accept(Event.CYCLE_FINISHED);
                }
                currentAnimation.start();
            }
        }

        @Override
        public void onAnimationCancel(Animator animation) {
            resetIndicatorPosition();
        }

        @Override
        public void onAnimationRepeat(Animator animation) {

        }
    };

    private void invokeStartCallback(Animator animation) {
        if (animation == down) {
            callback.accept(Event.DOWN);
        } else if (animation == right) {
            callback.accept(Event.RIGHT);
        } else if (animation == up) {
            callback.accept(Event.UP);
        } else if (animation == left) {
            callback.accept(Event.LEFT);
        }
    }

    private ObjectAnimator getNextAnimation(Animator animation) {
        if (animation == down) {
            return right;
        } else if (animation == right) {
            return up;
        } else if (animation == up) {
            return left;
        } else {
            return down;
        }
    }

    private boolean isLastPartOfTheCycle(Animator animation) {
        return animation == left;
    }

    private void resetIndicatorPosition() {
        indicatorView.setTranslationX(0f);
        indicatorView.setTranslationY(0f);
    }

    public IndicatorAnimation(Activity activity, Consumer<Event> callback) {
        this.callback = callback;
        ButterKnife.bind(this, activity);

        Resources resources = activity.getResources();
        float indicatorDiameter = (float)resources.getDimensionPixelSize(R.dimen.indicator_diameter);
        float metronomePadding = (float)resources.getDimensionPixelSize(R.dimen.metronome_padding) * 2;
        float longPath = metronomeView.getHeight() - indicatorDiameter - metronomePadding;
        float shortPath = metronomeView.getWidth() - indicatorDiameter - metronomePadding;

        down = ObjectAnimator.ofFloat(indicatorView, "translationY", 0f, longPath).setDuration(2000);
        down.addListener(animatorListener);

        right = ObjectAnimator.ofFloat(indicatorView, "translationX", 0f, shortPath).setDuration(1000);
        right.addListener(animatorListener);

        up = ObjectAnimator.ofFloat(indicatorView, "translationY", longPath, 0f).setDuration(2000);
        up.addListener(animatorListener);

        left = ObjectAnimator.ofFloat(indicatorView, "translationX", shortPath, 0f).setDuration(1000);
        left.addListener(animatorListener);
    }

    public void start() {
        animationRunning = true;
        currentAnimation = down;
        currentAnimation.start();
    }

    public void stop() {
        if (animationRunning) {
            animationRunning = false;
            currentAnimation.cancel();
        }
    }

    public enum Event {
        DOWN, RIGHT, UP, LEFT, CYCLE_FINISHED
    }



}

File: app/src/main/java/hu/kts/cmetronome/ui/workout/IndicatorAnimation.kt
package hu.kts.cmetronome.ui.workout

import android.animation.ObjectAnimator
import android.view.ViewTreeObserver
import hu.kts.cmetronome.repository.Settings
import javax.inject.Inject

/**
 * AnimatorSet is avoided because it behaved strange on cancel.
 */
class IndicatorAnimation @Inject constructor(private val fragment: WorkoutFragment, private val settings: Settings) {

    private var animationRunning: Boolean = false
    private var down: ObjectAnimator = ObjectAnimator.ofFloat(fragment.binding.indicatorView, "translationY", 0f, 0f)
    private var right: ObjectAnimator = ObjectAnimator.ofFloat(fragment.binding.indicatorView, "translationY", 0f, 0f)
    private var up: ObjectAnimator = ObjectAnimator.ofFloat(fragment.binding.indicatorView, "translationY", 0f, 0f)
    private var left: ObjectAnimator = ObjectAnimator.ofFloat(fragment.binding.indicatorView, "translationY", 0f, 0f)
    private var currentAnimation: ObjectAnimator? = null
    private var longPath: Float = 0F
    private var shortPath: Float = 0F

    init {
        fragment.binding.indicatorView.viewTreeObserver.addOnGlobalLayoutListener(
                object : ViewTreeObserver.OnGlobalLayoutListener {
                    override fun onGlobalLayout() {
                        fragment.binding.indicatorView.viewTreeObserver.removeOnGlobalLayoutListener(this)
                        init()
                    }
                })
    }

    private fun init() {
        longPath = fragment.binding.metronomeIndicatorPath.height.toFloat()
        shortPath = fragment.binding.metronomeIndicatorPath.width.toFloat()

        down = ObjectAnimator.ofFloat(fragment.binding.indicatorView, "translationY", 0f, longPath)

        right = ObjectAnimator.ofFloat(fragment.binding.indicatorView, "translationX", 0f, shortPath)

        up = ObjectAnimator.ofFloat(fragment.binding.indicatorView, "translationY", longPath, 0f)

        left = ObjectAnimator.ofFloat(fragment.binding.indicatorView, "translationX", shortPath, 0f)
        resetIndicatorPosition()
    }

    fun start(direction: Direction) {
        cancel()
        animationRunning = true
        when (direction) {
            Direction.DOWN -> moveIndicator(0f, 0f)
            Direction.RIGHT -> moveIndicator(0f, longPath)
            Direction.UP -> moveIndicator(shortPath, longPath)
            Direction.LEFT -> moveIndicator(shortPath, 0f)
        }
        currentAnimation = when (direction) {
            Direction.DOWN -> down.setDuration(settings.repDownTime)
            Direction.RIGHT -> right.setDuration(settings.repPauseDownTime)
            Direction.UP -> up.setDuration(settings.repUpTime)
            Direction.LEFT -> left.setDuration(settings.repPauseUpTime)
        }

        currentAnimation?.start()
    }

    fun stop() {
        cancel()
        resetIndicatorPosition()
    }

    private fun cancel() {
        if (animationRunning) {
            animationRunning = false
            currentAnimation?.cancel()
        }
    }

    private fun resetIndicatorPosition() {
        if (settings.repStartsWithUp) {
            moveIndicator(shortPath, longPath)
        } else {
            moveIndicator(0f, 0f)
        }
    }

    private fun moveIndicator(x: Float, y: Float) {
        fragment.binding.indicatorView.translationX = x
        fragment.binding.indicatorView.translationY = y
    }

    enum class Direction {
        DOWN, RIGHT, UP, LEFT
    }


}
